lib/demuxeres/gstdemuxeres.c

```diff
@@ -0,0 +1,721 @@
+gst_parse_stream_get_vcodec_from_caps (GstCaps * caps)
+{
+  GstDemuxerVideoCodec ret = DEMUXER_ES_VIDEO_CODEC_UNKNOWN;
+  GstStructure *s;
+  const gchar *name;
+
+  if (!caps)
+    return ret;
+
+  s = gst_caps_get_structure (caps, 0);
+  if (!s)
+    goto beach;
+
+  name = gst_structure_get_name (s);
+  if (!name)
+    goto beach;
+
+  if (!strcmp (name, "video/x-h264")) {
+    ret = DEMUXER_ES_VIDEO_CODEC_H264;
+  } else if (!strcmp (name, "video/x-h265")) {
+    ret = DEMUXER_ES_VIDEO_CODEC_H265;
+  }
+
+beach:
+  return ret;
+}
+
+static GstDemuxerAudioCodec
+gst_parse_stream_get_acodec_from_caps (GstCaps * caps)
+{
+  GstDemuxerAudioCodec ret = DEMUXER_ES_AUDIO_CODEC_UNKNOWN;
+  GstStructure *s;
+  const gchar *name;
+
+  if (!caps)
+    return ret;
+
+  s = gst_caps_get_structure (caps, 0);
+  if (!s)
+    goto beach;
+
+  name = gst_structure_get_name (s);
+  if (!name)
+    goto beach;
+
+  if (!strcmp (name, "audio/x-aac")) {
+    ret = DEMUXER_ES_AUDIO_CODEC_AAC;
+  }
+
+beach:
+  return ret;
+}
+
+void
+gst_parse_stream_teardown (GstDemuxerEStream * stream)
+{
+  switch (stream->type) {
+    case DEMUXER_ES_STREAM_TYPE_VIDEO:
+      g_free (stream->data.video.profile);
+      g_free (stream->data.video.level);
+      break;
+    default:
+      break;
+  }
+  g_free (stream->stream_id);
+  g_free (stream);
+}
+
+GstDemuxerEStream *
+gst_parse_stream_create (GstDemuxerES * demuxer, GstPad * pad)
+{
+  GstDemuxerEStream *stream;
+  GstPad *funnel_pad;
+  GstDemuxerESPrivate *priv = demuxer->priv;
+  GstCaps *caps = gst_pad_query_caps (pad, NULL);
+
+
+  if (!caps) {
+    GST_ERROR
+        ("Unable to get the caps from pad, unable to create a new stream");
+    return NULL;
+  }
+  GST_INFO ("the stream caps is %" GST_PTR_FORMAT, caps);
+
+  stream = g_new0 (GstDemuxerEStream, 1);
+  stream->demuxer = demuxer;
+  stream->type = gst_parse_stream_get_type_from_pad (pad);
+  stream->stream_id = gst_pad_get_stream_id (GST_PAD_CAST (pad));
+  GstStructure *s = gst_caps_get_structure (caps, 0);
+  switch (stream->type) {
+    case DEMUXER_ES_STREAM_TYPE_VIDEO:
+    {
+      gst_video_info_from_caps (&stream->data.video.info, caps);
+      gst_structure_get_int (s, "bitrate", &stream->data.video.bitrate);
+      stream->data.video.profile =
+          g_strdup (gst_structure_get_string (s, "profile"));
+      stream->data.video.level =
+          g_strdup (gst_structure_get_string (s, "level"));
+      stream->data.video.vcodec = gst_parse_stream_get_vcodec_from_caps (caps);
+      break;
+    }
+    case DEMUXER_ES_STREAM_TYPE_AUDIO:
+    {
+      gst_audio_info_from_caps (&stream->data.audio.info, caps);
+      gst_structure_get_int (s, "bitrate", &stream->data.audio.bitrate);
+      stream->data.audio.acodec = gst_parse_stream_get_acodec_from_caps (caps);
+      break;
+    }
+    default:
+      break;
+  }
+  stream->id = g_list_length (priv->streams);
+
+  funnel_pad = gst_element_request_pad_simple (priv->funnel, "sink_%u");
```

Comment:
From here, imo, it should handled in `uridecodebin_pad_added_cb()` since it's unrelated with the stream creation.